In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [301]:
df = pd.read_csv("../Data/Data_Train.csv")

In [302]:
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR ? DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU ? IXR ? BBI ? BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL ? LKO ? BOM ? COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU ? NAG ? BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR ? NAG ? DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [303]:
df.tail()

Airline Date_of_Journey    Source Destination  \
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time Arrival_Time Duration Total_Stops  \
10678              CCU ? BLR    19:55        22:25   2h 30m    non-stop   
10679              CCU ? BLR    20:45        23:20   2h 35m    non-stop   
10680              BLR ? DEL    08:20        11:20       3h    non-stop   
10681              BLR ? DEL    11:30        14:10   2h 40m    non-stop   
10682  DEL ? GOI ? BOM ? COK    10:55        19:15   8h 20m     2 stops   

      Additional_Info  Price  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753

In [304]:
df.shape

(10683, 11)

This dataset contain 10,683 rows and 11 columns.

In [305]:
df.duplicated().sum()

220

There are 220 duplicate values in that dataset, Now drop all the duplicate rows.

In [306]:
df.drop_duplicates(inplace = True)

In [307]:
df.shape

(10463, 11)

Now it contain 10,463 rows and 11 columns.

In [308]:
df.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

In [309]:
df.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [310]:
all_col = list(df.columns)
num_col = list(df._get_numeric_data())
cat_col = list(set(all_col) - set(num_col))

print("All features :", all_col)
print("\nNumerical features :", num_col)
print("\nCategorical features :", cat_col)

All features : ['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route', 'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops', 'Additional_Info', 'Price']

Numerical features : ['Price']

Categorical features : ['Additional_Info', 'Total_Stops', 'Dep_Time', 'Date_of_Journey', 'Airline', 'Source', 'Arrival_Time', 'Destination', 'Route', 'Duration']


In [311]:
df.nunique().sort_values(ascending = False)

Price              1870
Arrival_Time       1343
Duration            368
Dep_Time            222
Route               128
Date_of_Journey      44
Airline              12
Additional_Info      10
Destination           6
Source                5
Total_Stops           5
dtype: int64

In [312]:
df['Date_of_Journey'].value_counts()

6/06/2019     490
18/05/2019    486
9/06/2019     485
12/06/2019    483
21/05/2019    482
9/05/2019     466
21/03/2019    412
15/05/2019    402
27/05/2019    369
27/06/2019    339
24/06/2019    330
1/06/2019     330
3/06/2019     326
15/06/2019    314
24/03/2019    314
6/03/2019     302
27/03/2019    290
24/05/2019    286
6/05/2019     282
1/05/2019     274
12/05/2019    259
1/04/2019     256
3/03/2019     217
9/03/2019     199
15/03/2019    162
18/03/2019    156
01/03/2019    151
12/03/2019    141
9/04/2019     125
3/04/2019     110
21/06/2019    109
18/06/2019    105
09/03/2019    100
6/04/2019     100
06/03/2019     95
27/04/2019     94
24/04/2019     92
03/03/2019     92
3/05/2019      90
15/04/2019     89
21/04/2019     82
18/04/2019     67
12/04/2019     63
1/03/2019      47
Name: Date_of_Journey, dtype: int64

In [313]:
df['Jorney_day'] = df['Date_of_Journey'].str.split('/', expand = True)[0].astype('int')

In [314]:
df['Jorney_month'] = df['Date_of_Journey'].str.split('/', expand = True)[1].astype('int')

In [315]:
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR ? DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU ? IXR ? BBI ? BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL ? LKO ? BOM ? COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU ? NAG ? BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR ? NAG ? DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  \
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897   
1    05:50         13:15   7h 25m     2 stops         No info   7662   
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882   
3    18:05         23:30   5h 25m      1 stop         No info   6218   
4    16:50         21:35   4h 45m      1 stop         No info  13302   

   Jorney_day  Jorney_month  
0          24             3  
1           1             5  
2           9             6  
3          12             5  
4           1             3

In [316]:
df.drop(columns = 'Date_of_Journey', axis = 1, inplace = True)

In [317]:
df.head()

Airline    Source Destination                  Route Dep_Time  \
0       IndiGo  Banglore   New Delhi              BLR ? DEL    22:20   
1    Air India   Kolkata    Banglore  CCU ? IXR ? BBI ? BLR    05:50   
2  Jet Airways     Delhi      Cochin  DEL ? LKO ? BOM ? COK    09:25   
3       IndiGo   Kolkata    Banglore        CCU ? NAG ? BLR    18:05   
4       IndiGo  Banglore   New Delhi        BLR ? NAG ? DEL    16:50   

   Arrival_Time Duration Total_Stops Additional_Info  Price  Jorney_day  \
0  01:10 22 Mar   2h 50m    non-stop         No info   3897          24   
1         13:15   7h 25m     2 stops         No info   7662           1   
2  04:25 10 Jun      19h     2 stops         No info  13882           9   
3         23:30   5h 25m      1 stop         No info   6218          12   
4         21:35   4h 45m      1 stop         No info  13302           1   

   Jorney_month  
0             3  
1             5  
2             6  
3             5  
4             3

In [318]:
df['Jorney_day'].value_counts()

9     1375
6     1269
27    1092
21    1085
1     1058
24    1022
15     967
12     946
3      835
18     814
Name: Jorney_day, dtype: int64

In [319]:
df['Jorney_month'].value_counts()

5    3396
6    3311
3    2678
4    1078
Name: Jorney_month, dtype: int64

In [320]:
df['Dep_hour'] = pd.to_datetime(df['Dep_Time']).dt.hour
df['Dep_min'] = pd.to_datetime(df['Dep_Time']).dt.minute

df.drop(columns = ['Dep_Time'], axis = 1, inplace = True)

In [321]:
df['Arr_hour'] = pd.to_datetime(df['Arrival_Time']).dt.hour
df['Arr_min'] = pd.to_datetime(df['Arrival_Time']).dt.minute

df.drop(columns = ['Arrival_Time'], axis = 1, inplace = True)

In [322]:
df.head()

Airline    Source Destination                  Route Duration  \
0       IndiGo  Banglore   New Delhi              BLR ? DEL   2h 50m   
1    Air India   Kolkata    Banglore  CCU ? IXR ? BBI ? BLR   7h 25m   
2  Jet Airways     Delhi      Cochin  DEL ? LKO ? BOM ? COK      19h   
3       IndiGo   Kolkata    Banglore        CCU ? NAG ? BLR   5h 25m   
4       IndiGo  Banglore   New Delhi        BLR ? NAG ? DEL   4h 45m   

  Total_Stops Additional_Info  Price  Jorney_day  Jorney_month  Dep_hour  \
0    non-stop         No info   3897          24             3        22   
1     2 stops         No info   7662           1             5         5   
2     2 stops         No info  13882           9             6         9   
3      1 stop         No info   6218          12             5        18   
4      1 stop         No info  13302           1             3        16   

   Dep_min  Arr_hour  Arr_min  
0       20         1       10  
1       50        13       15  
2       25         4       25  
3        5        23       30  
4       50        21       35

In [323]:
df['hour'] = df['Duration'].str.split(expand = True)[0]
df['Min'] = df['Duration'].str.split(expand = True)[1]

In [324]:
df.head()

Airline    Source Destination                  Route Duration  \
0       IndiGo  Banglore   New Delhi              BLR ? DEL   2h 50m   
1    Air India   Kolkata    Banglore  CCU ? IXR ? BBI ? BLR   7h 25m   
2  Jet Airways     Delhi      Cochin  DEL ? LKO ? BOM ? COK      19h   
3       IndiGo   Kolkata    Banglore        CCU ? NAG ? BLR   5h 25m   
4       IndiGo  Banglore   New Delhi        BLR ? NAG ? DEL   4h 45m   

  Total_Stops Additional_Info  Price  Jorney_day  Jorney_month  Dep_hour  \
0    non-stop         No info   3897          24             3        22   
1     2 stops         No info   7662           1             5         5   
2     2 stops         No info  13882           9             6         9   
3      1 stop         No info   6218          12             5        18   
4      1 stop         No info  13302           1             3        16   

   Dep_min  Arr_hour  Arr_min hour   Min  
0       20         1       10   2h   50m  
1       50        13       15   7h   25m  
2       25         4       25  19h  None  
3        5        23       30   5h   25m  
4       50        21       35   4h   45m

In [325]:
df['hour'] = df['hour'].str.split('h', expand = True)[0]
df['Min'] = df['Min'].str.split('m', expand = True)[0]
df.drop(columns = ['Duration'], axis = 1, inplace = True)

In [326]:
print(df['hour'].unique())
print("\n")
print(df['Min'].unique())

['2' '7' '19' '5' '4' '15' '21' '25' '13' '12' '26' '22' '23' '20' '10'
 '6' '11' '8' '16' '3' '27' '1' '14' '9' '18' '17' '24' '30' '28' '29'
 '37' '34' '38' '35' '36' '47' '33' '32' '31' '42' '39' '5m' '41' '40']


['50' '25' None '45' '30' '5' '15' '35' '10' '20' '55' '40']


In [327]:
df['hour'] = df['hour'].str.split('m', expand = True)[0]
df['Min'] = df['Min'].astype('int', errors = 'ignore')

In [328]:
print(df['hour'].unique())
print("\n")
print(df['Min'].unique())

['2' '7' '19' '5' '4' '15' '21' '25' '13' '12' '26' '22' '23' '20' '10'
 '6' '11' '8' '16' '3' '27' '1' '14' '9' '18' '17' '24' '30' '28' '29'
 '37' '34' '38' '35' '36' '47' '33' '32' '31' '42' '39' '41' '40']


['50' '25' None '45' '30' '5' '15' '35' '10' '20' '55' '40']


In [329]:
df['Min'].isnull().sum()

1006

In [330]:
df['Min'].fillna('0', inplace = True)

In [331]:
print(df['hour'].unique())
print("\n")
print(df['Min'].unique())

['2' '7' '19' '5' '4' '15' '21' '25' '13' '12' '26' '22' '23' '20' '10'
 '6' '11' '8' '16' '3' '27' '1' '14' '9' '18' '17' '24' '30' '28' '29'
 '37' '34' '38' '35' '36' '47' '33' '32' '31' '42' '39' '41' '40']


['50' '25' '0' '45' '30' '5' '15' '35' '10' '20' '55' '40']


In [333]:
df['hour'] = df['hour'].astype('float')
df['Min'] = df['Min'].astype('float')

In [334]:
df['Duration(min)'] = df['hour'] * 60 + df['Min']

In [336]:
df.drop(columns = ['hour', 'Min'], axis = 1, inplace = True)

In [338]:
df.head()

Airline    Source Destination                  Route Total_Stops  \
0       IndiGo  Banglore   New Delhi              BLR ? DEL    non-stop   
1    Air India   Kolkata    Banglore  CCU ? IXR ? BBI ? BLR     2 stops   
2  Jet Airways     Delhi      Cochin  DEL ? LKO ? BOM ? COK     2 stops   
3       IndiGo   Kolkata    Banglore        CCU ? NAG ? BLR      1 stop   
4       IndiGo  Banglore   New Delhi        BLR ? NAG ? DEL      1 stop   

  Additional_Info  Price  Jorney_day  Jorney_month  Dep_hour  Dep_min  \
0         No info   3897          24             3        22       20   
1         No info   7662           1             5         5       50   
2         No info  13882           9             6         9       25   
3         No info   6218          12             5        18        5   
4         No info  13302           1             3        16       50   

   Arr_hour  Arr_min  Duration  
0         1       10     170.0  
1        13       15     445.0  
2         4       25    1140.0  
3        23       30     325.0  
4        21       35     285.0

In [339]:
all_col = list(df.columns)
num_col = list(df._get_numeric_data())
cat_col = list(set(all_col) - set(num_col))

print("All features :", all_col)
print("\nNumerical features :", num_col)
print("\nCategorical features :", cat_col)

All features : ['Airline', 'Source', 'Destination', 'Route', 'Total_Stops', 'Additional_Info', 'Price', 'Jorney_day', 'Jorney_month', 'Dep_hour', 'Dep_min', 'Arr_hour', 'Arr_min', 'Duration']

Numerical features : ['Price', 'Jorney_day', 'Jorney_month', 'Dep_hour', 'Dep_min', 'Arr_hour', 'Arr_min', 'Duration']

Categorical features : ['Additional_Info', 'Total_Stops', 'Airline', 'Source', 'Destination', 'Route']


In [340]:
for i in cat_col:
    print(i)
    print(df[i].value_counts())
    print("=" * 100)

Additional_Info
No info                         8183
In-flight meal not included     1926
No check-in baggage included     318
1 Long layover                    19
Change airports                    7
Business class                     4
No Info                            3
1 Short layover                    1
Red-eye flight                     1
2 Long layover                     1
Name: Additional_Info, dtype: int64
Total_Stops
1 stop      5625
non-stop    3475
2 stops     1318
3 stops       43
4 stops        1
Name: Total_Stops, dtype: int64
Airline
Jet Airways                          3700
IndiGo                               2043
Air India                            1695
Multiple carriers                    1196
SpiceJet                              815
Vistara                               478
Air Asia                              319
GoAir                                 194
Multiple carriers Premium economy      13
Jet Airways Business                    6
Vistara Premium econo